### PR TITLE
feat(model): add neural network framework

### PR DESCRIPTION
**Pull request by sahandsafizadeh**
- Feat!: complete ***qeep*** framework infrastructure by adding `model` tree.
using a *contract-based* approach, `model` wires up everything that already exists in `component` and `tensor` trees. the result, is a framework that allows us to define deep neural network in a rather *declarative* way.
the contracts are: (1) `Layer`, (2) `WeightedLayer`, (3) `Loss`, (4) `Metric`, (5) `Optimizer` and (6) `BatchGenerator`; each represented by an `interface`.
`node` is an internal `struct` that forms the core of `model`. each node represents a `Layer` and by creating bidirectional connections, nodes allow layers to pass computed results in a directed graph. thus, it provides a dynamic way to declare neural networks. in simpler terms, input tensors are *fed* to input nodes, network of nodes is traversed via ***BFS*** algorithm and finally, output nodes can be used for *prediction* and *optimization*.
defining network arechitecture in a declarative way, is done using `stream` package. it is inspired from `keras` *functional api* and gives simpler syntax with less burden of `error` handling. it gathers error of all layers during definition; and in the end, returns a total comprehensive per-layer error.
finally, having a valid `model.Model`, we can use it to `Predict` data or `Eval` its performance; or, we can `Fit` data and watch the process with the built-in `EpochLogger` that shows *duration* and *progress*.
- Feat: add built-in `Simple` batch-generator for *tabular* data.
it should be noted that both inputs and outputs are taken to be `2D` because of the nature of *batch* in neural networks. also, by design, all `components` and `models` in ***qeep*** framework take multiple input tensors and return one output tensor.
- Feat: add `examples` tree.
as the best documentation is code, this project is initially provided with three examples on benchmark datasets: (1) ***regression*** on `boston-housing`, (2) ***binary classification*** on `breast-cancer` and (3) ***multi-class classification*** on `iris`.

### CODE REVIEW

**Review by sahandsafizadeh — COMMENTED** [review]
reviewed

**Review by sahandsafizadeh — COMMENTED** [review]
reviewed
